mmpose/core/evaluation/top_down_eval.py

```diff
@@ -327,35 +404,56 @@ def keypoints_from_heatmaps(heatmaps,
         - maxvals (np.ndarray[N, K, 1]): Scores (confidence) of the keypoints.
     """
 
-    preds, maxvals = _get_max_preds(heatmaps)
     N, K, H, W = heatmaps.shape
-
-    if post_process:
-        if unbiased:  # alleviate biased coordinate
-            assert kernel > 0
-            # apply Gaussian distribution modulation.
-            heatmaps = _gaussian_blur(heatmaps, kernel)
-            heatmaps = np.maximum(heatmaps, 1e-10)
-            heatmaps = np.log(heatmaps)
-            for n in range(N):
-                for k in range(K):
-                    preds[n][k] = _taylor(heatmaps[n][k], preds[n][k])
-        else:
-            # add +/-0.25 shift to the predicted locations for higher acc.
-            for n in range(N):
-                for k in range(K):
-                    heatmap = heatmaps[n][k]
-                    px = int(preds[n][k][0])
-                    py = int(preds[n][k][1])
-                    if 1 < px < W - 1 and 1 < py < H - 1:
-                        diff = np.array([
-                            heatmap[py][px + 1] - heatmap[py][px - 1],
-                            heatmap[py + 1][px] - heatmap[py - 1][px]
-                        ])
-                        preds[n][k] += np.sign(diff) * .25
+    if use_udp:
+        assert target_type in ['GaussianHeatMap', 'CombinedTarget']
+        if target_type == 'GaussianHeatMap':
+            preds, maxvals = _get_max_preds(heatmaps)
+            if post_process:
+                preds = post_dark_udp(preds, heatmaps, kernel=kernel)
+        elif target_type == 'CombinedTarget':
+            for person_heatmaps in heatmaps:
+                for i, heatmap in enumerate(person_heatmaps):
+                    kt = 2 * kernel + 1 if i % 3 == 0 else kernel
+                    cv2.GaussianBlur(heatmap, (kt, kt), 0, heatmap)
```

Comment:
The heatmaps are smoothed inplace. Again, double check that it is safe.

Actually I want to suggest to copy the  heatmaps before this process, so that the function is polite and safe

Reply:
it is save here